### PR TITLE
Feature/acks naks and errorhandling (#122)

### PR DESCRIPTION
**Pull request by magraef**
As discussed in #122 , I have implemented an ackowladgement with Ack/Nak for the broker implementation and added an errorhandler for subscribers.

### CODE REVIEW

File: pkg/codegen/generators/v3/templates/controller.tmpl [hunk 1]
```diff
@@ -17,6 +17,7 @@ func New{{ .Prefix }}Controller(bc extensions.BrokerController, options ...Contr
         subscriptions:  make(map[string]extensions.BrokerChannelSubscription),
         logger:         extensions.DummyLogger{},
         middlewares:    make([]extensions.Middleware, 0),
+        errorHandler: errorhandlers.Noop(),
```
Comment (lerenn) [inline pkg/codegen/generators/v3/templates/controller.tmpl:20]: Same as last comment but for v3.

```suggestion
       errorHandler:   extensions.DefaultErrorHandler(),
```

File: pkg/extensions/brokers/natsjetstream/natsjetstream.go [hunk 2]
```diff
@@ -187,13 +199,23 @@ func (c *Controller) HandleMessage(msg jetstream.Msg, sub extensions.BrokerChann
...
+			Payload: msg.Data(),
+		},
+		AcknowledgementHandler{
+			doAck: func() {
+				if err := msg.Ack(); err != nil {
+					c.logger.Error(ctx, fmt.Sprintf("error on ack message: %q", err.Error()))
+				}
+			},
+			doNak: func() {
+				if err := msg.NakWithDelay(c.nakDelay); err != nil {
```
Comment (lerenn) [inline pkg/extensions/brokers/natsjetstream/natsjetstream.go:214]: Out of curiosity : is there any reason to use the `WithDelay`function ? :)

File: pkg/extensions/errorhandler.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package extensions
+
+import "context"
+
+// ErrorHandler is the signature of the function that needs to be implemented to use
+// errorhandler functionality
+type ErrorHandler func(ctx context.Context, topic string, msg *AcknowledgeableBrokerMessage, err error)
```
Comment (lerenn) [inline pkg/extensions/errorhandler.go:7]: Linked to the default handler comments.

```suggestion
type ErrorHandler func(ctx context.Context, topic string, msg *AcknowledgeableBrokerMessage, err error)

// DefaultErrorHandler returns the default error handler, which is Noop one.
func DefaultErrorHandler() ErrorHandler {
	return errorhandlers.Noop()
}
```

File: pkg/extensions/errorhandler.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package extensions
+
+import "context"
```
Comment (lerenn) [inline pkg/extensions/errorhandler.go:3]: Linked to default handler comments.

```suggestion
import (
	"context"
	
	"github.com/lerenn/asyncapi-codegen/pkg/extensions/errorhandlers"
)
```

**Review by lerenn — COMMENTED** [review]
Incredible work ! The code is neat and easy to read. Apart from a few comments which are more nitpicking than request for change, I have nothing more to add ! Thanks again for your work ☺️ 

There would be a few things to add like:
 - Change examples in the READMEs
 - Add tests for the error handling

But you've already done a lot with this PR, so if you don't have the time or the will to do it, please tell me and I'll do it :)

File: pkg/extensions/brokers/natsjetstream/natsjetstream.go (hunk 2 above)
Comment (magraef) [reply pkg/extensions/brokers/natsjetstream/natsjetstream.go:214]: In your draft you had described a nak with delay, so I included that. However, the nakDelay can also be set to 0 by the user.

File: pkg/codegen/generators/v2/templates/controller.tmpl [hunk 5]
```diff
@@ -17,6 +17,7 @@ func New{{ .Prefix }}Controller(bc extensions.BrokerController, options ...Contr
         subscriptions:  make(map[string]extensions.BrokerChannelSubscription),
         logger:         extensions.DummyLogger{},
         middlewares:    make([]extensions.Middleware, 0),
+		errorHandler:   errorhandlers.Noop(),
```
[the remark this answers is not in the document]
Comment (magraef) [reply pkg/codegen/generators/v2/templates/controller.tmpl:20]: yes makes perfect sense

File: examples/helloworld/v3/nats/app/main.go [hunk 6]
```diff
@@ -28,10 +28,13 @@ func main() {
 	// Subscribe to SayHelloMessage messages
 	// Note: it will indefinitely wait for messages as context has no timeout
 	log.Println("Subscribe to hello world...")
-	ctrl.SubscribeToReceiveHelloOperation(context.Background(), func(_ context.Context, msg SayHelloMessage) {
+	err = ctrl.SubscribeToReceiveHelloOperation(context.Background(), func(_ context.Context, msg SayHelloMessage) error {
 		log.Println("Received message:", msg.Payload)
+		return nil
 	})
-	defer ctrl.UnsubscribeFromReceiveHelloOperation(context.Background())
```
[the remark this answers is not in the document]
Comment (magraef) [reply examples/helloworld/v3/nats/app/main.go:34]: Good point. I think that was a mistake when rebasing.

**Comment by magraef** [comment]
@lerenn I will submit the tests and examples asap. Thanks for your review

**Comment by magraef** [comment]
@lerenn I have taken your advice on the last commits. The exception is the noop handler. I could not import it because of cycled imports. I have replaced it accordingly. Should now resolve #122 

I am unsure whether the tests are correctly located under issue 122. Also the setup of the tests is currently not documented or? Maybe you have a better idea for the tests.

**Comment by lerenn** [comment]
Thanks for the incredible work ! 💯 

Also the test is at the right place and with the correct setup, I will do a Contributing guide because it clearly lacks information on what should be done to contribute and, more specifically, how to test things.

I'll merge your MR ASAP ! Thanks again ☺️

**Comment by lerenn** [comment]
@magraef  the PR is merge !

Thanks again for your work, that will be really useful for a lot of people 🙏 

Feel free to continue contributing, either by code or by advice/critics ! I'll do a proper contributing guide in the next days :P

**Pull request merged**